### PR TITLE
[FN] Fix for transaction build verification

### PR DESCRIPTION
**Pull request by fenix2222**
@fassadlr Please review as per our discussion

### CODE REVIEW

**Comment by dangershony** [comment]
Can you eplainmore?

File: src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs [hunk 1]
```diff
@@ -56,8 +56,16 @@ public Transaction BuildTransaction(TransactionBuildContext context)
...
 
-            if (context.TransactionBuilder.Verify(transaction, out TransactionPolicyError[] errors))
-                return transaction;
+            bool validationSuccess = false;
+            TransactionPolicyError[] errors = null;
+
+            var retryOptions = new RetryOptions(3, TimeSpan.FromMilliseconds(5), RetryStrategyType.Simple, typeof(ArgumentNullException));
+            RetryStrategy.Run(retryOptions, () =>
+                {
+                    validationSuccess = context.TransactionBuilder.Verify(transaction, out errors);
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs]: @fenix2222 `BuildTransaction` should also be in the retry.

**Comment by fassadlr** [comment]
@dangershony NBitcoin's transaction builder does fee estimation by building a transaction with a (possible) random set of UTXOs and then discards that built transaction. If we then call BuildTransaction again to build and sign the *actual* transaction, a (possible) different set of random UTXOs could be used. This means that it is possible for the fee to not cover the selected set of UTXO's when we rebuild.

TLDR; NBitcoin's transaction builder does not use the same set of UTXO's to estimate the fee and build the transaction with.

The current version of NBitcoin's transaction builder does some retry strategy to mitigate this issue.

File: src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs [hunk 2]
```diff
@@ -54,9 +54,18 @@ public Transaction BuildTransaction(TransactionBuildContext context)
             if (context.Shuffle)
                 context.TransactionBuilder.Shuffle();
 
-            Transaction transaction = context.TransactionBuilder.BuildTransaction(context.Sign);
+            bool validationSuccess = false;
+            Transaction transaction = null;
+            TransactionPolicyError[] errors = null;
 
-            if (context.TransactionBuilder.Verify(transaction, out TransactionPolicyError[] errors))
+            var retryOptions = new RetryOptions(3, TimeSpan.FromMilliseconds(5), RetryStrategyType.Simple, typeof(ArgumentNullException));
```
Comment (MithrilMan) [inline src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs]: is delay time important here?

File: src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs [hunk 3]
```diff
@@ -54,9 +54,18 @@ public Transaction BuildTransaction(TransactionBuildContext context)
...
+            RetryStrategy.Run(retryOptions, () =>
+            {
+                transaction = context.TransactionBuilder.BuildTransaction(context.Sign);
+                validationSuccess = context.TransactionBuilder.Verify(transaction, out errors);
+            });
+
+            if (validationSuccess)
```
Comment (MithrilMan) [inline src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs]: so we retry without checking the error reason?
I think we should retry only if errors are "volatile" and a rerun may effectively generate a valid transaction

**Comment by MithrilMan** [comment]
@dangershony 
> Can you explain more?

Francois already replied, I just add a reference to a comment on a discussion we already had on this problem

https://github.com/stratisproject/StratisBitcoinFullNode/pull/3930#discussion_r323805314

**Comment by dangershony** [comment]
So TB is not deterministic? And this is a behaviour copied from current NBitcoin?

I wonder if a better approach is not possible.
What about capturing the utxo set once a TRX is built and removing all the rest unused UTXOs so the fee will always be the same?

**Comment by fassadlr** [comment]
@dangershony Yes, I believe NBitcoin does something similar at the moment:

https://github.com/MetacoSA/NBitcoin/blob/master/NBitcoin/TransactionBuilder.cs#L1374

and

https://github.com/MetacoSA/NBitcoin/blob/master/NBitcoin/TransactionBuilder.cs#L1425

I have tried to do as you suggested, to use the same set of UTXOs but it proved to hard a task to change the code now. It will be looked at again, just not at this present time.

File: src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs (hunk 2 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs]: @MithrilMan This is just part of the architecture at the moment. I suppose it is not that important but the method does not currently allow to not specify a delay unless you use the default constructor.

File: src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs (hunk 3 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs]: We will fix this 👍

**Comment by fassadlr** [comment]
@MithrilMan do you want to re-review this?

File: src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs [hunk 4]
```diff
@@ -55,20 +55,20 @@ public Transaction BuildTransaction(TransactionBuildContext context)
...
             int retryCount = 0;
 
             TransactionPolicyError[] errors = null;
             while (retryCount <= maxRetries)
             {
+                Transaction transaction = context.TransactionBuilder.BuildTransaction(false);
+                if (context.Sign)
+                {
+                    ICoin[] coinsSpent = context.TransactionBuilder.FindSpentCoins(transaction);
+                    this.AddSecrets(context, coinsSpent);
```
Comment (MithrilMan) [inline src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs]: We may have already added keys from the previous try, ensure that it works and doesn't rise exceptions.

Even if this works, this could be optimized to prevent the computation of keys we already computed in the previous try (because computing keys is one of the slowest part), but since this shouldn't happen so often you could leave as it is now, but add a TODO comment to flag this part as "needed to be improved"
